File: example_lib_common/src/main/java/com/example/common/utils/CompressUtil.java
package com.example.common.utils;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Matrix;
import android.media.ExifInterface;

import com.example.framework.utils.LogUtil;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;

/**
 * @author wyb
 * 图片压缩工具类
 */
public class CompressUtil {
    private static int defaultLength = 512;

    public static ByteArrayOutputStream compressImg(Bitmap image) {
        return compressImg(image, defaultLength);
    }

    public static ByteArrayOutputStream compressImg(Bitmap bitmap, int length) {
        bitmap = compressImgBySize(bitmap);
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        bitmap.compress(Bitmap.CompressFormat.JPEG, 100, byteArrayOutputStream);
        int options = 100;
        while (byteArrayOutputStream.toByteArray().length / 1024 > length) {
            byteArrayOutputStream.reset();
            options -= 10;
            bitmap.compress(Bitmap.CompressFormat.JPEG, options, byteArrayOutputStream);
        }
        LogUtil.INSTANCE.e("length", byteArrayOutputStream.toByteArray().length + "");
        return byteArrayOutputStream;
    }

    public static Bitmap compressImgBySize(Bitmap bitmap) {
        float size = 1f;
        int width = bitmap.getWidth();
        int height = bitmap.getHeight();
        Matrix matrix = new Matrix();
        if (width > 720) {
            size = 720f / width;
            LogUtil.INSTANCE.e("w", bitmap.getWidth() + "");
            LogUtil.INSTANCE.e("h", bitmap.getHeight() + "");
        } else if (height > 1280) {
            size = 1280f / height;
            LogUtil.INSTANCE.e("w", bitmap.getWidth() + "");
            LogUtil.INSTANCE.e("h", bitmap.getHeight() + "");
        }
        matrix.postScale(size, size);
        bitmap = Bitmap.createBitmap(bitmap, 0, 0, width, height, matrix, true);
        return bitmap;
    }

    public static File scale(Context context, File mFile) {
        final long fileMaxSize = 100 * 1024;
        return scale(context, mFile, fileMaxSize);
    }

    public static File scale(Context context, File mFile, long fileMaxSize) {
        long fileSize = mFile.length();
        float scaleSize = 1;
        LogUtil.INSTANCE.e("fileSize_old", mFile.length() / 1024 + "kb");
        if (fileSize >= fileMaxSize) {
            try {
                BitmapFactory.Options options = new BitmapFactory.Options();
                options.inJustDecodeBounds = true;
                Bitmap bitmap;
                int width = options.outWidth;
                int height = options.outHeight;
                if (width > 720) {
                    scaleSize = width / 480;
                } else if (height > 1280) {
                    scaleSize = width / 1080;
                }
                options.inJustDecodeBounds = false;
                options.inSampleSize = (int) (scaleSize + 0.5);
                bitmap = BitmapFactory.decodeFile(mFile.getPath(), options);
                bitmap = compressImgBySize(bitmap);

                File fTemp = new File(context.getApplicationContext().getExternalCacheDir(), System.currentTimeMillis() + "img.jpg");
                FileOutputStream fileOutputStream;
                try {
                    fileOutputStream = new FileOutputStream(fTemp);
                } catch (FileNotFoundException e) {
                    e.printStackTrace();
                    return mFile;
                }
                bitmap.compress(Bitmap.CompressFormat.JPEG, 100, fileOutputStream);

                fileOutputStream.flush();
                fileOutputStream.close();
                bitmap.recycle();
                LogUtil.INSTANCE.e("fileSize", fTemp.length() / 1024 + "kb");
                return fTemp;
            } catch (IOException e) {
                e.printStackTrace();
                return mFile;
            }
        } else {
            LogUtil.INSTANCE.e("fileSize", mFile.length() / 1024 + "kb");
            return mFile;
        }
    }

    //直接压缩图片
    public static Bitmap scale(Context context, Bitmap bitmap) {
        long fileSize = getBitmapSize(bitmap);
        float scaleSize = 1;
        final long fileMaxSize = 100 * 1024;
        if (fileSize >= fileMaxSize) {
            try {
                BitmapFactory.Options options = new BitmapFactory.Options();
                options.inJustDecodeBounds = true;
                int width = options.outWidth;
                int height = options.outHeight;
                if (width > 720) {
                    scaleSize = width / 480;
                } else if (height > 1280) {
                    scaleSize = width / 1080;
                }
                options.inJustDecodeBounds = false;
                options.inSampleSize = (int) (scaleSize + 0.5);
                bitmap = compressImgBySize(bitmap);

                File file = new File(context.getApplicationContext().getExternalCacheDir(), System.currentTimeMillis() + "img.jpg");
                FileOutputStream fileOutputStream;
                try {
                    fileOutputStream = new FileOutputStream(file);
                } catch (FileNotFoundException e) {
                    e.printStackTrace();
                    return bitmap;
                }
                bitmap.compress(Bitmap.CompressFormat.JPEG, 100, fileOutputStream);
                fileOutputStream.flush();
                fileOutputStream.close();
                return bitmap;
            } catch (IOException e) {
                return bitmap;
            }
        } else {
            return bitmap;
        }
    }

    public static int getBitmapSize(Bitmap bitmap) {
        return bitmap.getAllocationByteCount();
    }

    public static byte[] getBytes(InputStream inputStream) throws IOException {
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        byte[] buffer = new byte[1024]; // 用数据装
        int len;
        while ((len = inputStream.read(buffer)) != -1) {
            byteArrayOutputStream.write(buffer, 0, len);
        }
        byteArrayOutputStream.close();
        return byteArrayOutputStream.toByteArray();
    }

    public static File degreeImage(Context context, File mFile) {
        int degree = readImageDegree(mFile.getPath());
        Bitmap bitmap;
        if (degree != 0) {
            //旋转图片
            Matrix matrix = new Matrix();
            matrix.postRotate(degree);
            bitmap = BitmapFactory.decodeFile(mFile.getPath());
            bitmap = Bitmap.createBitmap(bitmap, 0, 0, bitmap.getWidth(), bitmap.getHeight(), matrix, true);
            File fTemp = new File(context.getApplicationContext().getExternalCacheDir(), System.currentTimeMillis() + "img.jpg");
            FileOutputStream fileOutputStream;
            try {
                fileOutputStream = new FileOutputStream(fTemp);
                bitmap.compress(Bitmap.CompressFormat.JPEG, 100, fileOutputStream);
                fileOutputStream.flush();
                fileOutputStream.close();
                bitmap.recycle();
                return fTemp;
            } catch (IOException e) {
                e.printStackTrace();
                return mFile;
            }
        } else {
            return mFile;
        }
    }

    //读取图片的方向
    private static int readImageDegree(String path) {
        int degree = 0;
        // 读取图片文件信息的类ExifInterface
        ExifInterface exifInterface = null;
        try {
            exifInterface = new ExifInterface(path);
        } catch (IOException e) {
            e.printStackTrace();
        }

        if (exifInterface != null) {
            int orientation = exifInterface.getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL);
            switch (orientation) {
                case ExifInterface.ORIENTATION_ROTATE_90:
                    degree = 90;
                    break;
                case ExifInterface.ORIENTATION_ROTATE_180:
                    degree = 180;
                    break;
                case ExifInterface.ORIENTATION_ROTATE_270:
                    degree = 270;
                    break;
            }
        }
        return degree;
    }

}

File: lib_base/src/main/java/com/dataqin/base/utils/CompressUtil.kt
package com.dataqin.base.utils

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.media.ExifInterface
import com.dataqin.base.utils.LogUtil.e
import java.io.*
import java.util.*


/**
 * Created by WangYanBin on 2020/7/23.
 * 图片工具类
 */
object CompressUtil {

    private fun log(msg: String) = LogUtil.e("CompressUtil", msg)

    /**
     * 压缩图片
     */
    @JvmOverloads
    @JvmStatic
    fun compress(bit: Bitmap, length: Int = 512): ByteArrayOutputStream {
        val bitTmp = compressBySize(bit)
        val byteArrayOutputStream = ByteArrayOutputStream()
        bitTmp.compress(Bitmap.CompressFormat.JPEG, 100, byteArrayOutputStream)
        var options = 100
        while (byteArrayOutputStream.toByteArray().size / 1024 > length) {
            byteArrayOutputStream.reset()
            options -= 10
            bitTmp.compress(Bitmap.CompressFormat.JPEG, options, byteArrayOutputStream)
        }
        log("图片")
        e("length", byteArrayOutputStream.toByteArray().size.toString())
        return byteArrayOutputStream
    }

    /**
     * 大小压缩
     */
    @JvmStatic
    private fun compressBySize(bit: Bitmap): Bitmap {
        var size = 1f
        val width = bit.width
        val height = bit.height
        val matrix = Matrix()
        if (width > 720) {
            size = 720f / width
            e("w", bit.width.toString())
            e("h", bit.height.toString())
        } else if (height > 1280) {
            size = 1280f / height
            e("w", bit.width.toString())
            e("h", bit.height.toString())
        }
        matrix.postScale(size, size)
        return Bitmap.createBitmap(bit, 0, 0, width, height, matrix, true)
    }

    /**
     * 质量压缩
     */
    @JvmStatic
    private fun scale(context: Context, file: File, fileMaxSize: Long = 100 * 1024): File {
        val fileSize = file.length()
        var scaleSize = 1f
        return if (fileSize >= fileMaxSize) {
            try {
                val options = BitmapFactory.Options()
                options.inJustDecodeBounds = true
                val width = options.outWidth
                val height = options.outHeight
                if (width > 720) {
                    scaleSize = width / 480f
                } else if (height > 1280) {
                    scaleSize = width / 1080f
                }
                options.inJustDecodeBounds = false
                options.inSampleSize = (scaleSize + 0.5).toInt()
                var bitmap = BitmapFactory.decodeFile(file.path, options)
                bitmap = compressBySize(bitmap)
                val tempFile = File(
                    context.getExternalFilesDir(null)?.absolutePath,
                    "${DateUtil.getDateTime("yyyyMMdd_HHmmss", Date())}.jpg"
                )
                val fileOutputStream = try {
                    FileOutputStream(tempFile)
                } catch (e: FileNotFoundException) {
                    return file
                }
                bitmap.compress(Bitmap.CompressFormat.JPEG, 100, fileOutputStream)
                fileOutputStream.flush()
                fileOutputStream.close()
                bitmap.recycle()
                tempFile
            } catch (e: IOException) {
                file
            }
        } else file
    }

    /**
     * 质量压缩
     */
    @JvmStatic
    private fun scale(context: Context, bitmap: Bitmap, fileMaxSize: Long = 100 * 1024): Bitmap {
        var bitmapTmp = bitmap
        val fileSize = bitmapTmp.allocationByteCount
        var scaleSize = 1f
        return if (fileSize >= fileMaxSize) {
            try {
                val options = BitmapFactory.Options()
                options.inJustDecodeBounds = true
                val width = options.outWidth
                val height = options.outHeight
                if (width > 720) {
                    scaleSize = width / 480f
                } else if (height > 1280) {
                    scaleSize = width / 1080f
                }
                options.inJustDecodeBounds = false
                options.inSampleSize = (scaleSize + 0.5).toInt()
                bitmapTmp = compressBySize(bitmapTmp)
                val file = File(
                    context.getExternalFilesDir(null)?.absolutePath,
                    "${DateUtil.getDateTime("yyyyMMdd_HHmmss", Date())}.jpg"
                )
                val fileOutputStream = try {
                    FileOutputStream(file)
                } catch (e: FileNotFoundException) {
                    return bitmapTmp
                }
                bitmapTmp.compress(Bitmap.CompressFormat.JPEG, 100, fileOutputStream)
                fileOutputStream.flush()
                fileOutputStream.close()
                bitmapTmp
            } catch (e: IOException) {
                bitmapTmp
            }
        } else bitmapTmp
    }

    /**
     * 旋转图片
     */
    @JvmStatic
    fun degree(context: Context, file: File): File {
        val degree = readDegree(file.path)
        var bitmap: Bitmap
        return if (degree != 0f) {
            //旋转图片
            val matrix = Matrix()
            matrix.postRotate(degree)
            bitmap = BitmapFactory.decodeFile(file.path)
            bitmap = Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)
            val tempFile = File(
                context.getExternalFilesDir(null)?.absolutePath,
                DateUtil.getDateTime("yyyyMMdd_HHmmss", Date()) + ".jpg"
            )
            val fileOutputStream: FileOutputStream
            try {
                fileOutputStream = FileOutputStream(tempFile)
                bitmap.compress(Bitmap.CompressFormat.JPEG, 100, fileOutputStream)
                fileOutputStream.flush()
                fileOutputStream.close()
                bitmap.recycle()
                tempFile
            } catch (e: IOException) {
                file
            }
        } else file
    }

    /**
     * 读取图片的方向
     */
    @JvmStatic
    fun readDegree(path: String): Float {
        var degree = 0f
        //读取图片文件信息的类ExifInterface
        var exifInterface: ExifInterface? = null
        try {
            exifInterface = ExifInterface(path)
        } catch (ignored: IOException) {
        } finally {
            when (exifInterface?.getAttributeInt(
                ExifInterface.TAG_ORIENTATION,
                ExifInterface.ORIENTATION_NORMAL
            )) {
                ExifInterface.ORIENTATION_ROTATE_90 -> degree = 90f
                ExifInterface.ORIENTATION_ROTATE_180 -> degree = 180f
                ExifInterface.ORIENTATION_ROTATE_270 -> degree = 270f
            }
        }
        return degree
    }

    /**
     * 旋转图片
     */
    @JvmStatic
    fun degree(angle: Int, bitmap: Bitmap): Bitmap {
        var returnBm: Bitmap? = null
        // 根据旋转角度，生成旋转矩阵
        val matrix = Matrix()
        matrix.postRotate(-angle.toFloat())
        try {
            // 将原始图片按照旋转矩阵进行旋转，并得到新的图片
            returnBm = Bitmap.createBitmap(
                bitmap,
                0,
                0,
                bitmap.width,
                bitmap.height,
                matrix,
                true
            )
        } catch (e: OutOfMemoryError) {
        }
        if (returnBm == null) {
            returnBm = bitmap
        }
        if (bitmap != returnBm) {
            bitmap.recycle()
        }
        return returnBm
    }

}